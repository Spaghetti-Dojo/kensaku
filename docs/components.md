---
title: Components
layout: post
nav_order: 2
---

# Components

The Wp Entities Search provides a set of components you can use to build your search UI.

We have to distinguish between two types of components:

- **Base Components**: These are the basic building blocks of the search UI. They are the smallest components and are
  used to build more complex components.
- **Composite Components**: These are more complex components that are built using the base ones.

The _Composite Component_ does not have a direct dependency on the _Base Components_, indeed a composite component
encapsulate the logic
that deal with the state management of the data and pass it down to the base components in form of properties via
the `children` function.

In a nutshell, a _Composite Component_ is configured from the external via props, dialog with the Rest API and pass the
data down to the _Base Components_.

## Base Components

- `PluralSelectControl` - A select control that allows multiple selections.
- `SingleSelectControl` - A select control that allows a single selection.
- `SearchControl` - An input control that allows the user to enter a search query.
- `ToggleControl` - A multi selection control that allows the user to toggle between multiple options.
- `RadioControl` - A single selection control that allows the user to select a single option.

## Composite Components

- `CompositeEntitiesByKind` - A composite component that displays a list of entities by kind. In this context _kind_
  is `post` or `term` and _entity_ is the entity type of the content e.g. a `page` or a `category` term.

## Composite Entities by Kind

The `CompositeEntitiesByKind` is a _generic_ component that can be used to display a list of entities by kind. It acts
as a controller that fetches the data from the Rest API using a given function and pass it down to one or more _Base
Components_.
It is up to you how to consume the data given to the children. Note, even if you consume just one part of the given
data,
the component will fetch all the data necessary to render the UI.

This component is intended to fulfill a scenario where the user is looking for some entities belonging to a certain
_kind_,
for instance, selecting the `page` _kind_ will return a list of the available pages and to filter the result further it
is
possible to use a search function in conjunction with a `search` field.

An example of its usage is:

```jsx
import { CompositeEntitiesByKind } from 'wp-entities-search';

export function MyComponent(props) {
    const entities = {
        value: new Set([13, 24, 55]),
        onChange: (value) => {
            // Do something with the selected entities
            // Notice, here you'll receive the value as a string Set.
        }
    };

    const kind = {
        value: new Set(['page']),
        options: [
            { label: 'Pages', value: 'page' },
            { label: 'Posts', value: 'post' },
        ],
        onChange={(value) => {
            // Do something with the selected kind
        }}
    };

    const searchEntities = async (phrase, kind, queryArguments) => convertEntitiesToControlOptions(
        await searchEntities('post', kind, phrase, queryArguments),
        'title',
        'id'
    );

    return <CompositeEntitiesByKind
        searchEntities={searchEntities}
        entities={entities}
        kind={kind}
    >
        {(entities, kind, search) => (
            <>
                <ToggleControl
                    value={kind.value}
                    options={kind.options}
                    onChange={kind.onChange}
                />
                <SearchControl
                    onChange={search.onChange}
                />
                <ToggleControl
                    value={entities.value}
                    onChange={entities.onChange}
                />
            </>
        )}
    </CompositeEntitiesByKind>;
}
```

In the example above we are passing a `searchEntities` function that will be used to fetch the entities from the Rest
API. This function is documented in the [api](./api.md) section.

What's important to know is that the `queryArguments` do not include the `_fields` property because it's not a concern
of the component decide which fields to retrieve, it's up to the consumer to decide which fields to retrieve. Anyhow,
you can still override the given configuration, but you have to be aware this might produce an incorrect result if not
done correctly. The component might pass arguments necessary to guarantee the consistency of the Set of the entities and
the kind.

In the following example we require `title` and `slug` fields to be retrieved from the Rest API.

```typescript
 async (phrase, kind, queryArguments) => {
    const fields = ['title', 'slug'];
    return convertEntitiesToControlOptions(
        await searchEntities('post', kind, phrase, {...queryArguments, fields}),
        ...fields
    )
 }
```

The `entities` and `kind` properties are the initial configuration, they'll change depending on what happen within the
respective `onChange` functions.

The `children` function will receive the `entities`, `kind` and `search` properties; notice the `search` is a function,
and it's up to the `SearchControl` to maintain the status of the value, the composite does hold the search value
internally, but it does not share it outside.

You're not forced to consume all the properties given to the `children`, you can remove the `SearchControl`
and let the user select only the retrieved entities. Moreover, you can also not have the Kind control at all and just
allow the search for one single kind.

In the example below we only allow to select the entities belonging to the `page` or `landing-page` kind not permitting
the user to switch between them.

```jsx
import { CompositeEntitiesByKind } from 'wp-entities-search';

export function MyComponent(props) {
    const entities = {
        value: new Set([13, 24, 55]),
        onChange: (value) => {
            // Do something with the selected entities
        }
    };

    const kind = {
        value: new Set(['page', 'landing-page']),
        options: Set([]),
        onChange={() => {}}
    };

    const searchEntities = async (phrase, kind, queryArguments) => convertEntitiesToControlOptions(
        await searchEntities('post', kind, phrase, queryArguments),
        'title',
        'id'
    );

    return <CompositeEntitiesByKind
        searchEntities={searchEntities}
        entities={entities}
        kind={kind}
    >
        {(entities, _, search) => (
            <>
                <SearchControl
                    onChange={search.onChange}
                />
                <ToggleControl
                    value={entities.value}
                    options={entities.options}
                    onChange={entities.onChange}
                />
            </>
        )}
    </CompositeEntitiesByKind>;
}
```

Obviously depending on what you want to achieve you can use different _Base Components_ or create new ones, as mentioned
above the package comes with a set of _Base Components_ that can be used out of the box.

## About Singular Base Components

The _Composite Component_ always give a collection of Entities and Kind even though you are consuming a Single* _Base
Component_.

The Singular Components always get a single value, therefore you have to consider to extract the first element out of
the `Set`.

```jsx
import { CompositeEntitiesByKind } from 'wp-entities-search';

export function MyComponent(props) {
    return <CompositeEntitiesByKind
        /* ... */
    >
        {(entities, kind, search) => (
            <>
                <RadioControl
                    value={kind.value.first()}
                    options={kind.options}
                    onChange={(value) => kind.onChange(new Set([value]))}
                />
                /* ... */
            </>
        )}
    </CompositeEntitiesByKind>;
}
```
