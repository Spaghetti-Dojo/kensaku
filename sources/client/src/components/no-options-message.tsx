import React, { JSX } from 'react';

import { __ } from '@wordpress/i18n';

export function NoOptionsMessage(): JSX.Element {
	return (
		<p className="wes-no-option-message">
			{__('No options', 'wp-entities-search')}
		</p>
	);
}
