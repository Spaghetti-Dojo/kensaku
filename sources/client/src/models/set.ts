/**
 * External dependencies
 */
import { isEqual as _isEqual } from 'lodash';

export class Set< T > {
	readonly #data: ReadonlyArray< T >;

	public constructor( data: ReadonlyArray< T > = [] ) {
		this.#data = this.ensureUnique( data );
	}

	public add( value: T ): Set< T > {
		if ( this.has( value ) ) {
			return this;
		}

		return new Set( [ ...this.#data, value ] );
	}

	public delete( value: T ): Set< T > {
		if ( ! this.has( value ) ) {
			return this;
		}

		return new Set(
			this.#data.filter( ( item ) => ! this.isEqual( item, value ) )
		);
	}

	public has( value: T ): boolean {
		return this._has( value, this.#data );
	}

	public map< R = T >( fn: ( value: T ) => R ): Set< R > {
		return new Set( this.#data.map( fn ) );
	}

	public toArray(): ReadonlyArray< T > {
		return Object.freeze( [ ...this.#data ] );
	}

	public forEach( fn: ( value: T ) => void ): void {
		this.#data.forEach( fn );
	}

	public length(): number {
		return this.#data.length;
	}

	public concat( set: Set< T > ): Set< T > {
		return new Set( [ ...this.#data, ...set.toArray() ] );
	}

	public filter( fn: ( value: T ) => boolean ): Set< T > {
		return new Set( this.#data.filter( fn ) );
	}

	public find( fn: ( value: T ) => boolean ): T | undefined {
		return this.#data.slice( 0 ).find( fn );
	}

	public first(): T | undefined {
		return this.#data.slice( 0 )[ 0 ];
	}

	public last(): T | undefined {
		return this.#data.slice( -1 )[ 0 ];
	}

	public copy( start: number, end: number ): Set< T > {
		return new Set( this.#data.slice( start, end ) );
	}

	public equals( set: Set< T > ): boolean {
		if ( this.length() !== set.length() ) {
			return false;
		}

		if ( this === set ) {
			return true;
		}

		for ( const value of this ) {
			if ( ! set.has( value ) ) {
				return false;
			}
		}

		return true;
	}

	public *[ Symbol.iterator ]() {
		for ( const value of this.#data ) {
			yield value;
		}
	}

	private isEqual( a: unknown, b: unknown ): boolean {
		return _isEqual( a, b );
	}

	private _has( value: T, data: ReadonlyArray< T > ): boolean {
		return data.some( ( current ) => this.isEqual( current, value ) );
	}

	private ensureUnique( data: ReadonlyArray< T > ): ReadonlyArray< T > {
		const accumulator: Array< T > = [];
		return data.reduce( ( acc, value ) => {
			if ( ! this._has( value, acc ) ) {
				acc.push( value );
			}

			return acc;
		}, accumulator );
	}
}
