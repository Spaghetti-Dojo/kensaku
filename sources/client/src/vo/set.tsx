export class Set<T> {
	readonly #data: ReadonlyArray<T>;

	public constructor(data: Array<T> = []) {
		this.#data = data;
	}

	public add(value: T): Set<T> {
		if (this.has(value)) {
			return this;
		}

		return new Set([...this.#data, value]);
	}

	public delete(value: T): Set<T> {
		if (!this.has(value)) {
			return this;
		}

		return new Set(this.#data.filter((item) => item !== value));
	}

	public has(value: T): boolean {
		return this.#data.includes(value);
	}

	public map<R = T>(fn: (value: T) => R): Set<R> {
		return new Set(this.#data.map(fn));
	}

	public toArray(): Array<T> {
		return [...this.#data];
	}

	public forEach(fn: (value: T) => void): void {
		this.#data.forEach(fn);
	}

	public length(): number {
		return this.#data.length;
	}

	public concat(set: Set<T>): Set<T> {
		return new Set([...this.#data, ...set.toArray()]);
	}

	public filter(fn: (value: T) => boolean): Set<T> {
		return new Set(this.#data.filter(fn));
	}

	public first(): T | undefined {
		return this.#data.slice(0)[0];
	}

	public last(): T | undefined {
		return this.#data.slice(-1)[0];
	}

	public copy(start: number, end: number): Set<T> {
		return new Set(this.#data.slice(start, end));
	}

	public equals(set: Set<T>): boolean {
		if (this.length() !== set.length()) {
			return false;
		}

		if (this === set) {
			return true;
		}

		for (const value of this) {
			if (!set.has(value)) {
				return false;
			}
		}

		return true;
	}

	public *[Symbol.iterator]() {
		for (const value of this.#data) {
			yield value;
		}
	}
}
