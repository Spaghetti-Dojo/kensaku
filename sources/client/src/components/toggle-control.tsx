/**
 * External dependencies
 */
import EntitiesSearch from '@types';
import classnames from 'classnames';
import React, { JSX } from 'react';

import { useId } from '../hooks/use-id';
import { NoOptionsMessage } from './no-options-message';

interface Option<V> extends EntitiesSearch.ControlOption<V> {
	readonly selectedValues: EntitiesSearch.BaseControl<V>['value'];
	readonly onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

export function ToggleControl(
	props: EntitiesSearch.BaseControl< EntitiesSearch.Value > & {
		className?: string;
	}
): JSX.Element {
	const className = classnames( props.className, 'wes-toggle-control' );

	if ( props.options.length() <= 0 ) {
		return <NoOptionsMessage />;
	}

	const onChange = ( event: React.ChangeEvent< HTMLInputElement > ) => {
		const { target } = event;
		const valueOption = props.options.find(
			( option ) => String( option.value ) === target.value
		);

		if ( ! valueOption ) {
			return;
		}

		if ( target.checked ) {
			props.onChange( props.value.add( valueOption.value ) );
			return;
		}

		props.onChange( props.value.delete( valueOption.value ) );
	};

	return (
		<div className={className}>
			{props.options.map((option) => (
				<Option<EntitiesSearch.Value>
					key={option.value}
					label={option.label}
					value={option.value}
					selectedValues={props.value}
					onChange={onChange}
				/>
			))}
		</div>
	);
}

function Option<V>(props: Option<V>): JSX.Element {
	const id = useId();
	const value = String(props.value);

	return (
		<div
			className={`wes-toggle-control-item wes-toggle-control-item--${value}`}
		>
			<label htmlFor={id}>
				<input
					type="checkbox"
					id={id}
					className={`wes-toggle-control-item__input-${value}`}
					checked={props.selectedValues?.has(props.value)}
					value={value}
					onChange={props.onChange}
				/>
				{props.label}
			</label>
		</div>
	);
}
