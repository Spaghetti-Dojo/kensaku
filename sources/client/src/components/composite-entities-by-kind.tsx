/**
 * External dependencies
 */
import type EntitiesSearch from '@types';
import React, { JSX } from 'react';

/**
 * WordPress dependencies
 */
import { doAction } from '@wordpress/hooks';

/**
 * Internal dependencies
 */
import { useEntitiesOptionsStorage } from '../hooks/use-entities-options-storage';
import { useSearch } from '../hooks/use-search';
import { orderSelectedOptionsAtTheTop } from '../utils/order-selected-options-at-the-top';
import { uniqueControlOptions } from '../utils/unique-control-options';
import { Set } from '../models/set';

/**
 * A composite component that provides a way to search for entities by kind.
 *
 * @public
 * @param props The component props.
 */
export function CompositeEntitiesByKind< E, K >(
	props: EntitiesSearch.CompositeEntitiesKinds< E, K >
): JSX.Element {
	const [ state, dispatch ] = useEntitiesOptionsStorage< E, K >(
		{
			entities: props.entities.value,
			kind: props.kind.value,
		},
		props.searchEntities
	);
	const search = useSearch< E, K >(
		props.searchEntities,
		state.kind,
		state.entities,
		dispatch
	);

	const onChangeEntities = ( entities: EntitiesSearch.Entities< E > ) => {
		props.entities.onChange( entities );

		if ( entities.length() <= 0 ) {
			dispatch( {
				type: 'UPDATE_SELECTED_ENTITIES_OPTIONS',
				selectedEntitiesOptions: new Set(),
			} );
			return;
		}

		Promise.all( [
			props.searchEntities( state.searchPhrase, state.kind, {
				exclude: entities,
			} ),
			props.searchEntities( '', state.kind, {
				include: entities,
				per_page: '-1',
			} ),
		] )
			.then( ( result ) => {
				const currentEntitiesOptions = result[ 0 ] ?? new Set();
				const selectedEntitiesOptions = result[ 1 ] ?? new Set();

				dispatch( {
					type: 'UPDATE_SELECTED_ENTITIES_OPTIONS',
					selectedEntitiesOptions,
				} );
				dispatch( {
					type: 'UPDATE_CURRENT_ENTITIES_OPTIONS',
					currentEntitiesOptions,
				} );
			} )
			.catch( ( error ) => {
				doAction(
					'wp-entities-search.on-change-entities.error',
					error
				);
			} );
	};

	const onChangeKind = ( kind: EntitiesSearch.Kind< K > ) => {
		const emptySet = new Set< any >();

		props.kind.onChange( kind );
		props.entities.onChange( emptySet );

		if ( kind.length() <= 0 ) {
			dispatch( {
				type: 'CLEAN_ENTITIES_OPTIONS',
			} );
			dispatch( { type: 'UPDATE_KIND', kind } );
			return;
		}

		props
			.searchEntities( state.searchPhrase, kind, {
				exclude: state.entities,
			} )
			.then( ( entitiesOptions ) => {
				dispatch( {
					type: 'UPDATE_ENTITIES_OPTIONS_FOR_NEW_KIND',
					entitiesOptions,
					kind,
				} );
			} )
			.catch( ( error ) => {
				dispatch( {
					type: 'CLEAN_ENTITIES_OPTIONS',
				} );
				doAction( 'wp-entities-search.on-change-kind.error', error );
			} );
	};

	const entities: EntitiesSearch.BaseControl< E > = {
		...props.entities,
		value: state.entities,
		options: orderSelectedOptionsAtTheTop< E >(
			uniqueControlOptions(
				state.selectedEntitiesOptions.concat(
					state.currentEntitiesOptions
				)
			),
			state.entities
		),
		onChange: onChangeEntities,
	};

	const kind: EntitiesSearch.BaseControl< K > = {
		...props.kind,
		value: state.kind,
		onChange: onChangeKind,
	};

	return <>{ props.children( entities, kind, search ) }</>;
}

// A Comment
