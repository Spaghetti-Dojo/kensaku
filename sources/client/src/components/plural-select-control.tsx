import EntitiesSearch from '@types';
import classnames from 'classnames';
import React, { JSX } from 'react';

import { Set } from '../vo/set';
import { NoOptionsMessage } from './no-options-message';

export function PluralSelectControl(
	props: EntitiesSearch.BaseControl<string> & {
		className?: string;
	}
): JSX.Element {
	const [selected, setSelected] = React.useState(props.value);
	const className = classnames(
		props.className,
		'wz-select-control',
		'wz-select-control--plural'
	);

	if (props.options.length() <= 0) {
		return <NoOptionsMessage />;
	}

	return (
		<select
			multiple
			className={className}
			value={selected.toArray()}
			onChange={(event) => {
				if (event.target.selectedOptions.length <= 0) {
					props.onChange(new Set());
				}

				const selectedValues = Array.from(
					event.target.selectedOptions
				).map((option) => option.value);

				setSelected(new Set(selectedValues));
				props.onChange(new Set(selectedValues));
			}}
		>
			{props.options.map((option) => (
				<option
					key={option.value}
					className={`wz-select-control-item wz-select-control-item--${option.value}`}
					value={option.value}
				>
					{option.label}
				</option>
			))}
		</select>
	);
}
