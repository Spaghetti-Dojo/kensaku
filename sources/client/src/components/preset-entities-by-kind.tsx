/**
 * External dependencies
 */
import EntitiesSearch from '@types';
import React, { JSX } from 'react';
import classnames from 'classnames';

/**
 * WordPress dependencies
 */
import { createHigherOrderComponent } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import { CompositeEntitiesByKind } from './composite-entities-by-kind';
import { SearchControl } from './search-control';
import { Set } from '../models/set';

type EntitiesValue = EntitiesSearch.Value;
type Entities = EntitiesSearch.Entities< EntitiesValue >;
type KindValue = EntitiesSearch.Kind< string > | string;

type EntitiesFinder = (
	phrase: string,
	kind: EntitiesSearch.Kind< string >,
	queryArguments?: EntitiesSearch.QueryArguments
) => Promise< Set< EntitiesSearch.ControlOption< EntitiesValue > > >;

type EntitiesComponent = React.ComponentType<
	EntitiesSearch.BaseControl< EntitiesValue >
>;
type KindComponent = React.ComponentType<
	EntitiesSearch.BaseControl< KindValue >
>;

type PublicComponentProps = {
	entitiesFinder: EntitiesFinder;
	className?: string;
	entities: Entities;
	onChangeEntities: ( values: Entities ) => void;
	entitiesComponent: EntitiesComponent;
	kind: KindValue;
	kindOptions: EntitiesSearch.Options< string >;
	onChangeKind: ( values: KindValue ) => void;
	kindComponent: KindComponent;
	entitiesFields?: EntitiesSearch.QueryArguments[ 'fields' ];
};

interface PrivateComponentProps
	extends Pick<
		EntitiesSearch.CompositeEntitiesKinds< EntitiesValue, string >,
		'kind' | 'entities'
	> {
	className?: string;
	entitiesFinder: EntitiesFinder;
	kindComponent: KindComponent;
	entitiesComponent: EntitiesComponent;
}

function PrivateComponent( props: PrivateComponentProps ): JSX.Element {
	const className = classnames( 'wes-preset-entities-by-kind', {
		// @ts-ignore
		[ props.className ]: !! props.className,
	} );

	return (
		<div className={ className }>
			<CompositeEntitiesByKind
				entities={ props.entities }
				kind={ props.kind }
				searchEntities={ props.entitiesFinder }
			>
				{ ( _entities, _kind, search ) => (
					<>
						<props.kindComponent { ..._kind } />
						<SearchControl onChange={ search } />
						<props.entitiesComponent { ..._entities } />
					</>
				) }
			</CompositeEntitiesByKind>
		</div>
	);
}

const withDataBound = createHigherOrderComponent<
	React.ComponentType< PrivateComponentProps >,
	React.ComponentType< PublicComponentProps >
>(
	( Component ) => ( props ) => {
		const {
			kind,
			entities,
			onChangeEntities,
			kindOptions,
			onChangeKind,
			entitiesFields,
			kindComponent,
			entitiesComponent,
			entitiesFinder,
			..._props
		} = props;

		const kindValue = narrowKindValue( props.kind );

		const _entities = {
			value: entities,
			onChange: onChangeEntities,
		};

		const _kind = {
			value: kindValue,
			options: kindOptions,
			onChange: onChangeKind,
		};

		const _entitiesFinder = entitiesFinderWithExtraFields(
			entitiesFinder,
			entitiesFields
		);

		return (
			<Component
				entities={ _entities }
				kind={ _kind }
				entitiesFinder={ _entitiesFinder }
				kindComponent={ kindComponent }
				entitiesComponent={ entitiesComponent }
				{ ..._props }
			/>
		);
	},
	'withDataBound'
);

function entitiesFinderWithExtraFields(
	entitiesFinder: EntitiesFinder,
	entitiesFields?: EntitiesSearch.QueryArguments[ 'fields' ]
): EntitiesFinder {
	return (
		phrase: string,
		kind: EntitiesSearch.Kind< string >,
		queryArguments?: EntitiesSearch.QueryArguments
	) =>
		entitiesFinder( phrase, kind, {
			...queryArguments,
			fields: [
				...( queryArguments?.fields ?? [ 'title', 'id' ] ),
				...( entitiesFields ?? [] ),
			],
		} );
}

function narrowKindValue( value: KindValue ): EntitiesSearch.Kind< string > {
	return typeof value === 'string' ? new Set( [ value ] ) : value;
}

export const PresetEntitiesByKind = withDataBound( PrivateComponent );
