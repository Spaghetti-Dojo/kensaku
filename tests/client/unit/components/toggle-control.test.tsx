/**
 * External dependencies
 */
import React from 'react';

import { describe, expect, it, jest } from '@jest/globals';

import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Internal dependencies
 */
import { ToggleControl } from '../../../../sources/client/src/components/toggle-control';
import { Set } from '../../../../sources/client/src/models/set';

const options = new Set( [
	{ label: 'Option 1', value: '1' },
	{ label: 'Option 2', value: '2' },
	{ label: 'Option 3', value: '3' },
] );

describe( 'EntitiesToggleControl', () => {
	it( 'renders correctly', () => {
		const { container } = render(
			<ToggleControl
				className="test-class"
				value={ new Set( [ '1' ] ) }
				options={ options }
				onChange={ () => {} }
			/>
		);

		expect( container.firstChild ).toMatchSnapshot();
	} );

	it( 'renders the NoOptionsMessage when there are no options', () => {
		const { container } = render(
			<ToggleControl
				className="test-class"
				value={ new Set( [ '1' ] ) }
				options={ new Set() }
				onChange={ () => {} }
			/>
		);

		expect( container.firstChild ).toMatchSnapshot();
	} );

	it( 'updates the value when one or more option are selected', () => {
		const onChange = jest.fn();

		render(
			<ToggleControl
				className="test-class"
				options={ options }
				value={ new Set( [] ) }
				onChange={ onChange }
			/>
		);

		fireEvent.click( screen.getByLabelText( 'Option 1' ) );

		expect( onChange ).toHaveBeenCalledWith( new Set( [ '1' ] ) );
	} );

	it( 'updates the value when one or more option are unselected', () => {
		const onChange = jest.fn();

		render(
			<ToggleControl
				className="test-class"
				options={ options }
				value={ new Set( [ '1', '2' ] ) }
				onChange={ onChange }
			/>
		);

		fireEvent.click( screen.getByLabelText( 'Option 1' ) );

		expect( onChange ).toHaveBeenCalledWith( new Set( [ '2' ] ) );
	} );

	it( 'does not change the value when an option is selected that does not exist', async () => {
		const props = {
			options: new Set( [
				{
					label: 'Option 1',
					value: 'option-one',
				},
				{
					label: 'Option 2',
					value: 'option-two',
				},
			] ),
			value: new Set( [ 'option-one' ] ),
			onChange: jest.fn(),
		};
		const rendered = render( <ToggleControl { ...props } /> );

		const option = rendered.container.querySelector(
			'.kensaku-toggle-control-item__input-option-one'
		) as HTMLOptionElement;
		option.value = 'option-3';

		await userEvent.click( option );

		expect( props.onChange ).not.toHaveBeenCalled();
	} );
} );
