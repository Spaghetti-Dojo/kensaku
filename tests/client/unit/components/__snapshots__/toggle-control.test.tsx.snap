// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitiesToggleControl renders correctly 1`] = `
<div
  class="test-class wz-toggle-control"
>
  <div
    class="wz-toggle-control-item wz-toggle-control-item--1"
  >
    <label
      for="wz-toggle-control-item__input-option-1-1"
    >
      <input
        checked=""
        class="wz-toggle-control-item__input-1"
        id="wz-toggle-control-item__input-option-1-1"
        type="checkbox"
        value="1"
      />
      Option 1
    </label>
  </div>
  <div
    class="wz-toggle-control-item wz-toggle-control-item--2"
  >
    <label
      for="wz-toggle-control-item__input-option-2-2"
    >
      <input
        class="wz-toggle-control-item__input-2"
        id="wz-toggle-control-item__input-option-2-2"
        type="checkbox"
        value="2"
      />
      Option 2
    </label>
  </div>
  <div
    class="wz-toggle-control-item wz-toggle-control-item--3"
  >
    <label
      for="wz-toggle-control-item__input-option-3-3"
    >
      <input
        class="wz-toggle-control-item__input-3"
        id="wz-toggle-control-item__input-option-3-3"
        type="checkbox"
        value="3"
      />
      Option 3
    </label>
  </div>
</div>
`;

exports[`EntitiesToggleControl renders the NoOptionsMessage when there are no options 1`] = `
<p
  class="wp-entities-search-no-option-message"
>
  No options
</p>
`;
